src/audio/mixin_mixout.c

```diff
@@ -758,8 +758,9 @@ static int mixout_reset(struct comp_dev *dev)
 			/* FIXME: this is racy and implicitly protected by serialised IPCs */
 			source = container_of(blist, struct comp_buffer, sink_list);
 			source_c = buffer_acquire(source);
-			stop = (dev->pipeline == source_c->source->pipeline &&
-					source_c->source->state > COMP_STATE_PAUSED);
+			stop = (source_c->source &&
+				dev->pipeline == source_c->source->pipeline &&
+				source_c->source->state > COMP_STATE_PAUSED);
```

Comment:
Not following the feedback @lyakh. All I did was to avoid a dereference. Are you saying we need to change the result stored in the 'stop' variable if the source is not defined? This is far from self-explanatory, sorry.

Reply:
what I mean is that the check for a `NULL` pointer can be done in multiple ways, e.g. either as an error or an ignore. E.g. I also recently made a similar mistake, in a loop
```
list_for_item(item, list) {
	object = list_item(item,...);
	if (!object->pointer)
		return -EINVAL;
}
```
i.e. `NULL` is an error, whereas the correct handling was
```
	if (!object->pointer)
		continue;
```
i.e. `NULL` is ignore.
Similar in your case: you calculate a `stop` variable which is then used to decide whether to stop looping and return `PPL_STATUS_PATH_STOP` immediately or to continue the loop. So, it's correct, that you shouldn't dereference `NULL`, but what should you do if you encounter one? You have the choice here: to continue the loop to the next source buffer or return `PPL_STATUS_PATH_STOP` immediately. You can expand it as:
```
if (!source_c->source)
	stop = true;
else
	stop = source_c->source &&
		dev->pipeline == source_c->source->pipeline &&
		source_c->source->state > COMP_STATE_PAUSED;
```
or the second line above could be `stop = false;` - both avoid `NULL` dereference but results are different.

Reply:
and BTW - the parentheses in
```
stop = (source_c->source &&
	dev->pipeline == source_c->source->pipeline &&
	source_c->source->state > COMP_STATE_PAUSED);
```
are superfluous just like they were in the original version

Reply:
@lyakh and @lgirdwood I don't have enough understanding of the firmware buffer management to have an opinion.

When a source buffer is invalid, does it makes sense to stop resetting the downstream components?
I didn't create the problem, the existing code is broken already

Reply:
@plbossart I'm not 100% sure myself, but in this specific case we seem to be making a decision, whether to propagate a `RESET` command or not, based on whether any sources are active. We make a decision to stop propagation if we find any active sources. So, yes, I think if `source_c->source == NULL` that means, that the respective source is inactive and we should continue the loop. So the implementation seems correct to me (apart from the parentheses), but the commit message should be fixed.

Reply:
ok, thanks @lyakh I will fix the commit message.